### PR TITLE
added a funtion that log kernel build time

### PR DESCRIPTION
**Pull request by PraveenPenguin**
Signed-off-by: Praveen K Pandey <praveen@linux.vnet.ibm.com>

1- Added a function which log `kernel build time`
2- added a function which do `make clean ()`

### CODE REVIEW

**Comment by PraveenPenguin** [comment]
@ldoktor can you please have a look in this PR 
Regards
Praveen

File: avocado/utils/kernel.py [hunk 1]
```diff
@@ -97,14 +97,40 @@ def build(self):
...
-            build.make(self.linux_dir, extra_args='O=%s olddefconfig' % self.build_dir)
+            build.make(self.linux_dir, extra_args='O=%s olddefconfig' %
+                       self.build_dir)
         build.make(self.linux_dir, extra_args='O=%s' % self.build_dir)
 
     def __del__(self):
         shutil.rmtree(self.work_dir)
 
+    def kernel_build_time(self, threads, timefile='/dev/null', make_opts='',
+                          output='/dev/null'):
```
Comment (sathnaga) [inline avocado/utils/kernel.py]: @PraveenPenguin How about adding this as an option for in build.make() method itself, https://github.com/avocado-framework/avocado/blob/master/avocado/utils/build.py#L21, that would make it more generic.
```

def make(path, make='make', env=None, extra_args='', ignore_status=False, allow_output_check='none', measure_time =False):
....
....
if measure_time:
       ......
        cmd = "time %s" % cmd
        .....
        .....

```

Some thing like this?

**Comment by PraveenPenguin** [comment]
Satheesh thanks for looking into this 
let me check but not sure because this depend on kernel build  , before that we need to download and extract in build_dir .in `kernel.py` those functionality already there , 

But this make sense we can add a functionality  for get build time for other build .
@ldoktor @apahim can you please share yours thought .

Regards
Praveen

**Comment by apahim** [comment]
`build.make()` uses `process.system()`.
`process.system()` is a wrapper to `process.run()` and only returns the exit code. On the other hand, `process.run()` returns also the `duration`:
```python
>>> from avocado.utils import process
>>> result = process.run('sleep 1')
>>> print result.duration
1.00145888329
>>> 
```
My suggestion is to change `build.make()` to use `process.run()`. It's a one line change.
Then the tests can have:
```python
...
result = build.make(...)
self.log.info(result.duration)
...
```

**Comment by PraveenPenguin** [comment]
@apahim  thanks looks me good suggestion , will try to do that .

**Comment by ldoktor** [comment]
Actually, @apahim, I can see a benefit in using the `/usr/bin/time` as it provides other useful information like user/system times, utilization, memory info...

@PraveenPenguin as for the style, do you expect to use this in many tests? To me this looks more like a stress test, then utils API, therefor I'd implement it inside the test itself. There are, though, ways to improve `utils.kernel` to support you in that. I'd suggest adding `build_cmd` method to `kernel.KernelBuild` (which requires to add `build_cmd` to the `build` library). Then in the test I'd do all steps to build the kernel, only before executing the actual build I'd call `cmd = kernel_build.build_cmd`, add the `time` and redirections to the command and run it as `process.system`.

@sathnaga Having it as a `build` option is an option, but I don't think it'd be used frequently (I mean by many tests). I see it as a special case so I'd rather see it as a separate entity (ideally sharing the common basis)

**Comment by ldoktor** [comment]
Also regarding the `process.system` vs `process.run`, yes, it'd help with the basic information, the only problem is that changing the API (return types) is quite hard and expensive (there are tests out there depending on the values). For a period of time we'd have to implement either a `variable` to report the new return, or a different method and slowly obsolete the other.

**Comment by ldoktor** [comment]
Last comment, don't take me for granted, I'm not sure what is your primary usage, requirements etc. So feel free to disagree with me and hopefully we'll reach the best solution.

**Comment by PraveenPenguin** [comment]
@ldoktor 

yes i want this kernel build time  functionality  in more than two test case , kernbench and kernel build  that's why raise this PR ,
not very sure as you say it is difficult to adopt  process.system instead of process.run but seems me good idea from @apahim 
till now we can use different function that use process.system like overloading and later we can obsolete this , @apahim @ldoktor  @sathnaga if all agree will try to implement this 

Regards
Praveen

**Comment by apahim** [comment]
@ldoktor yes, there are plenty of ways to change the api (you provided 2) or even create new api  like `buil.run_make()` or something. I see the requirement here as **log kernel build time**, so the process.run.duration can provide that. Ayway, I agree `build_cmd` can provide even more flexibility (and also the possibility to break the build cmd command). No strong opinion here.
Notice that some tests are also needing the output of the build.make() to check for errors and they are doing some juggle to achieve that (see https://github.com/PraveenPenguin/avocado-misc-tests/blob/050d40f2733c328591a1a7f8c10be130bfac7ecd/memory/libhugetlbfs.py#L124-L125), so process.run.stdout would be useful there as well.

**Comment by ldoktor** [comment]
Yep, if the duration is all what is needed, then adding the `run_make` is IMO the best way.

**Comment by PraveenPenguin** [comment]
@ldoktor @apahim can i add run_make method in build as overloaded method  . please help me

**Comment by apahim** [comment]
@PraveenPenguin seems like we came to an agreement and you can propose a `build.run_make()` method based on `build.make()` but using `process.run()` under the hood.

**Comment by PraveenPenguin** [comment]
@apahim  thanks for helping on this . will do

**Comment by sathnaga** [comment]
@PraveenPenguin  build.make() and process.run() would be a nice approach, @apahim @ldoktor Tnx :-)

**Comment by apahim** [comment]
Ping

**Comment by apahim** [comment]
Release freeze date is approaching fast. If you want to give this a chance to be included in next release, please send a new version.

**Comment by PraveenPenguin** [comment]
@apahim sure will do thanks

**Comment by PraveenPenguin** [comment]
@apahim @ldoktor raised a new PR #1449 please have look.

Regards
Praveen

**Pull request closed without merging**